### PR TITLE
[`unnecessary_find_map`]: look for then_some

### LINKED ISSUE #11260: `unnecessary_find_map` should see through `then_some`
### Summary

It would be nice if Clippy could also warn me about using `then_some` in return of a `find_map` closure, like in this example:

```
std::iter::once(1).find_map(|n| (n > 1).then_some(n));
```

### Lint Name

unnecessary_find_map

### Reproducer

I tried this code:

```rust
    // warns
    std::iter::once(1).find_map(|n| if n > 1 { Some(n) } else { None });
    // does not
    std::iter::once(1).find_map(|n| (n > 1).then_some(n));
    // better solution
    std::iter::once(1).find(|n| *n > 1);
```

I expected to see this happen:

The `unnecessary_find_map` lint should flag the second example as well

Instead, this happened:

No warning is emitted

### Version

_No response_

### PR DESCRIPTION
**Pull request by y21**
Closes #11260

changelog: [`unnecessary_find_map`]: lint `.then_some()` in closure

### CODE REVIEW

File: tests/ui/unnecessary_find_map.rs [hunk 1]
```diff
@@ -21,3 +21,7 @@ fn main() {
 fn find_map_none_changes_item_type() -> Option<bool> {
     "".chars().find_map(|_| None)
 }
+
+fn issue11260() {
+    let _x = std::iter::once(1).find_map(|n| (n > 1).then_some(n));
```
Comment (blyxyas) [inline tests/ui/unnecessary_find_map.rs:26]: Could you add some more tests? :cowboy_hat_face:

**Review by blyxyas — COMMENTED** [review]
Just a mini-nit, everything else seems pretty good :cow2:

File: tests/ui/unnecessary_find_map.rs (hunk 1 above)
Comment (y21) [reply tests/ui/unnecessary_find_map.rs:26]: Do you have anything specific in mind? I realized that this fix also applies to `unnecessary_filter_map`, so I'm going to add a test that it also does the right thing there, but I'm not sure what else we would want to test here

Comment (y21) [reply tests/ui/unnecessary_find_map.rs:27]: Added a test for unnecessary_filter_map and testing that it doesn't incorrectly suggest `.find()` when calling `.then_some()` with a different option

**Review by blyxyas — APPROVED** [review]
LGTM, Thanks :heart:!

**Pull request merged**